### ISSUE TITLE
Nested partials cause stack overflow

### ISSUE THREAD

**Issue by emi2k01** [sunng87/handlebars-rust#487]
```rs
use handlebars::Handlebars;

fn main() {
    let mut hb = Handlebars::new();
    hb.register_template_string("partial", "{{> @partial-block}}").unwrap();
    hb.register_template_string("index",
r#"{{#>partial}}
    Yo
    {{#>partial}}
    Yo 2
    {{/partial}}
{{/partial}}"#).unwrap();
    dbg!(hb.get_templates());

    hb.render("index", &()).unwrap();
}
```

The infinite recursion seems to be something like (top-down/outer-inner order):

- `Template::render` (1)
    - `TemplateElement::render` 
        - `partial::expand_partial`
            - `partial::find_partial`
                - `Registry::get_or_load_template_optional`
            - Back to (1)

**Comment by sunng87** [comment]
handlebarsjs renders this as

```
    Yo
    Yo 2
```

This could have something to do with how we store `@partial-block`

**Comment by emi2k01** [comment]
I thought it only crashed if the partial was the same but it also crashes with:

```handlebars
{{> partial}}
{{> partial2}}
:(
{{/partial2}}
{{/partial}}
```

My project's blocked on this so I'll try to fix it asap. Do you have any suggestions for where I should look at? I'll try to make sense of the call-stack I posted in the issue.

**Comment by emi2k01** [comment]
It seems that when the inner partial is rendering, the code that gets the template from `@partial-block` returns the `@partial-block` of the outer partial which makes it start rendering the outer partial again.

**Comment by emi2k01** [comment]
I think I found it.

`push_partial_block` prepends the partial block to the stack but `get_partial` always gets the partial block at the end of the queue.

So with:

```handlebars
{{#> partial1}}
{{#> partial2}}
{{#> partial3}}
{{/partial3}
{{/partial2}
{{/partial1}
```

nested partials the stack looks like `[partial 3, partial2, partial1]` and `get_partial` always gets `partial1`

**Comment by emi2k01** [comment]
I'm not sure what would be the best way to fix this.

**Comment by sunng87** [comment]
it seems `partial_block_depth` is not correctly maintained so `get_partial` failed to return its current one. investigating...

**Comment by sunng87** [comment]
@emi2k01 can you please check the fix in #488 if it works for your cases?

**Comment by emi2k01** [comment]
I think it does. Thank you so much
